wallet/src/de/schildbach/wallet/ui/widget/ShortcutsPane.kt

```diff
@@ -79,87 +81,96 @@ class ShortcutsPane(context: Context, attrs: AttributeSet) : LinearLayout(contex
+    private val shortcuts = listOf(
+        secureNowButton,
+        scanToPayButton,
+        payToContactButton,
+        buySellButton,
+        receiveButton,
+        explore
+    )
 
-    private var onShortcutClickListener: OnClickListener? = null
+    var isPassphraseVerified: Boolean = true
+        set(value) {
+            secureNowButton.shouldAppear = !value
+
+            if (field != value) {
+                field = value
+                refresh()
+            }
+        }
+
+    var userHasBalance: Boolean = true
+        set(value) {
+            scanToPayButton.shouldAppear = value
+            buySellButton.shouldAppear = !value
+            payToAddressButton.shouldAppear = value
+            payToContactButton.shouldAppear = value
+
+            if (field != value) {
+                field = value
+                refresh()
+            }
+        }
 
     init {
         setBackgroundResource(R.drawable.white_background_rounded)
         minimumHeight = 180
         orientation = HORIZONTAL
         gravity = Gravity.CENTER_HORIZONTAL
         if (isInEditMode) {
-            setup()
             refresh()
         }
         val onPreDrawListener = object : ViewTreeObserver.OnPreDrawListener {
             override fun onPreDraw(): Boolean {
                 viewTreeObserver.removeOnPreDrawListener(this)
                 val sizeRation = width.toFloat() / height.toFloat()
                 isSmallScreen = (sizeRation < 3.3)
-                setup()
                 refresh()
                 return false
             }
         }
         viewTreeObserver.addOnPreDrawListener(onPreDrawListener)
     }
 
-    fun setup() {
-        addShortcut(secureNowButton)
-        secondaryItems.add(scanToPayButton)
-        if (isSmallScreen) {
-            secondaryItems.add(receiveButton)
-        }
-        secondaryItems.add(payToContactButton)
-        secondaryItems.add(buySellButton)
-        secondaryItems.forEach {
-            addShortcut(it)
-        }
-        if (!isSmallScreen) {
-            addShortcut(receiveButton)
-        }
-    }
-
     private fun refresh() {
-        val displayed = mutableSetOf<ShortcutButton>()
-        secureNowButton.visibility = if (showSecureNow) {
-            displayed.add(secureNowButton)
-            View.VISIBLE
-        } else View.GONE
-        if (!isSmallScreen) {
-            displayed.add(receiveButton)
-        }
-        val numberOfButtons = if (isSmallScreen) 3 else 4
-        secondaryItems.forEach {
-            it.visibility = if (displayed.size < numberOfButtons) {
-                displayed.add(it)
-                View.VISIBLE
-            } else View.GONE
+        var slotsLeft = if (isSmallScreen) 3 else 4
+        shortcuts.forEach { btn ->
+            if (btn.shouldAppear && slotsLeft > 0) {
+                addShortcut(btn)
+                slotsLeft--
+            } else {
+                removeShortcut(btn)
+            }
```

Comment:
After trying to sort out our rules for displaying shortcuts, I've realized that there are 2 problems:
- Visibility of shortcuts is handled with both `addView` and `isVisible`. This leads to having 2 different methods `setup` and `refresh` while there should be only 1.
- `ShorcutsPane` mixes together the logic for each button and the logic for small/big screens. This should be split up.

The change I've made:
- Having buttons added/removed only by `addShortcut` and `removeShortcut` methods. There is no ambiguity as to how to show and hide those buttons anymore.
- Having all buttons we want to ever display in the current app version in the `shortcuts` list. This is the only place that dictates the order of buttons.
- The `shouldAppear` property in the `ShortcutButton` is the only thing that tells whether the conditions are satisfied for showing the button.
- `refresh` checks the list of buttons, checks which buttons want to be displayed and takes maximum based on the screen size.